experimental/devices/ads1x15/ads1x15.go

```diff
@@ -0,0 +1,392 @@
+		return
+	}
+	if err = d.checkChannel(channelB); err != nil {
+		return
+	}
+
+	if channelA == Channel0 && channelB == Channel1 {
+		mux = 0
+	} else if channelA == Channel0 && channelB == Channel3 {
+		mux = 1
+	} else if channelA == Channel1 && channelB == Channel3 {
+		mux = 2
+	} else if channelA == Channel2 && channelB == Channel3 {
+		mux = 3
+	} else {
+		err = errors.New("Only some differences of channels are allowed:  0 - 1, 0 - 3, 1 - 3 or 2 - 3")
+		return
+	}
+
+	return d.prepareQuery(mux, maxVoltage, minimumFrequency)
+}
+
+func (d *Dev) prepareQuery(mux int, maxVoltage physic.ElectricPotential, minimumFrequency physic.Frequency) (pin AnalogPin, err error) {
+	// Determine the most appropriate gain
+	gain, err := d.bestGainForElectricPotential(maxVoltage)
+	if err != nil {
+		return
+	}
+
+	// Validate the gain.
+	gainConf, ok := d.gainConfig[gain]
+	if !ok {
+		err = errors.New("Gain must be one of: 2/3, 1, 2, 4, 8, 16")
+		return
+	}
+
+	// Determine the voltage multiplier for this gain
+	voltageMultiplier, ok := d.gainVoltage[gain]
+	if !ok {
+		err = errors.New("Gain must be one of: 2/3, 1, 2, 4, 8, 16")
+		return
+	}
+
+	// Determine the most appropriate data rate
+	dataRate, err := d.bestDataRateForFrequency(minimumFrequency)
+	if err != nil {
+		return
+	}
+
+	dataRateConf, ok := d.dataRates[dataRate]
+
+	if !ok {
+		// Write a nice error message in case the data rate is not found
+		keys := []int{}
+		for k := range d.dataRates {
+			keys = append(keys, k)
+		}
+
+		err = fmt.Errorf("Invalid data rate. Accepted values: %d", keys)
+		return
+	}
+
+	// Build the configuration value
+	var config uint16
+	config = ads1x15ConfigOsSingle // Go out of power-down mode for conversion.
+	// Specify mux value.
+	config |= uint16((mux & 0x07) << ads1x15ConfigMuxOffset)
+	// Validate the passed in gain and then set it in the config.
+	config |= gainConf
+	// Set the mode (continuous or single shot).
+	config |= ads1x15ConfigModeSingle
+
+	// Set the data rate (this is controlled by the subclass as it differs
+	// between ADS1015 and ADS1115).
+	config |= dataRateConf
+	config |= ads1x15ConfigCompQueDisable // Disable comparator mode.
+
+	// Build the query to the ADC
+	configBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(configBytes, config)
+	preparedQuery := append([]byte{ads1x15PointerConfig}, configBytes...)
+
+	// The wait for the ADC sample to finish is based on the sample rate plus a
+	// small offset to be sure (0.1 millisecond).
+	waitTime := time.Duration(1000000/dataRate+100) * time.Microsecond
+
+	pin = &ads1x15AnalogPin{
+		adc:               d,
+		preparedQuery:     preparedQuery,
+		voltageMultiplier: voltageMultiplier,
+		waitTime:          waitTime,
+	}
+
+	return
+}
+
+func (d *Dev) executePreparedQuery(query []byte, waitTime time.Duration, voltageMultiplier physic.ElectricPotential) (value physic.ElectricPotential, err error) {
+	// Lock the ADC converter to avoid multiple simultaneous readings.
```

Comment:
Actually, are you sure it's a good idea? I mean maybe someone wants to read from multiple channels from different goroutines?

As a follow up: having a way to read all values at once may be useful for the user.
As a follow up: https://periph.io/x/periph/conn/physic#SenseEnv SenseContinuous is a nice way to have a stable continuous sampling, where the driver can optimize the configured sampling rate based on the frequency the user requested. This has to be stoppable via Halt().

Reply:
Unfortunately, those ADS1x15 devices are made of 1 ADC bundled with a multiplexer. So it's actually not possible to measure multiple channels simultaneously ([diagram](http://www.ti.com/general/docs/datasheetdiagram.tsp?genericPartNumber=ADS1115&diagramId=SBAS444D)).
If you want to do that, you need to have 2 physical ADC components (it's a good thing it's easy to change the I2C address to have them chained).

Follow-up #1: that's interesting, but we have to keep in mind that each pin can have different rate/gain settings. But we could imagine having a simplified version for that case, with shared settings. Do you have a particular use-case in mind?

Follow-up #2: actually, the ADS1x15 have a continuous mode where the ADC never sleeps and samples continuously at one of the fixed rates. However, when using that mode, it's not possible to use multiple pins, as the multiplexer is in a fixed state. So doing that in software could be a better idea, as it would be more flexible and would avoid having the user deal too much with the internals of the device. I'll try to do it, as it totally makes sense from a user perspective.